### PR TITLE
feat: enable the POST method for the `oidc.callback` route

### PR DESCRIPTION
**Pull request by eviweb**
This addresses the issue #3.

Signed-off-by: Eric Villard <dev@eviweb.fr>

### CODE REVIEW

**Comment by maicol07** [comment]
Maybe it can be helpful to add a note in the readme regarding CSFR

**Comment by eviweb** [comment]
@maicol07,
I try to find a better way to disable CSRF middleware, as the callback route is customizable using the `OIDC_CALLBACK_ROUTE_PATH` variable, and the `VerifyCsrfToken::$except` does not allowed to use `config('oidc.callback_route_path')`.
If you have any ideas, feel free to share ;).

**Comment by eviweb** [comment]
@maicol07,
maybe an idea to look into, could we use the `state` parameter to keep the CSRF token between API exchanges.
I think that's what it's for, right?
If we can then reuse it, it would be completely transparent to the end user.
What do you think about that?

**Comment by eviweb** [comment]
Hi @maicol07,

I think I get a proper solution, completely transparent for end users who want to use the `POST` method for their `callback` route.
Please give it a look at this POC: [https://github.com/eviweb/laravel-oidc-client/tree/fix-csrf](https://github.com/eviweb/laravel-oidc-client/tree/fix-csrf) (the last two commits).
The idea is to save the CSRF token in the `state` parameter of the OIDC request.
Then to use a middleware which will be called before the `VerifyCsrfToken` one, in order to retrieve the token value from the state parameter and rebuild the session token.

It works, but it implies to modify the initialization of the OIDC `state` parameter in the oidc-client-php package, here:
[https://github.com/maicol07/oidc-client-php/blob/4d16b7be3dc01e4915f373979a1726435ca76be4/src/Client.php#L320](https://github.com/maicol07/oidc-client-php/blob/4d16b7be3dc01e4915f373979a1726435ca76be4/src/Client.php#L320) by `$state = csrf_token();`.

What do you think about that?

**Comment by maicol07** [comment]
That's good. However, I think it's better to "lock" the middleware under a setting: add a setting in `config.php` that toggles the middleware registration. Also, you should restrict the middleware only for the return route, not for all the `web` requests: https://laravelpackage.com/11-middleware.html#route-middleware

> It works, but it implies to modify the initialization of the OIDC `state` parameter in the oidc-client-php package, here:  
> [maicol07/oidc-client-php@`4d16b7b`/src/Client.php#L320](https://github.com/maicol07/oidc-client-php/blob/4d16b7be3dc01e4915f373979a1726435ca76be4/src/Client.php#L320) by `$state = csrf_token();`.

Yes, I'll add a parameter to allow to set a custom state

**Comment by eviweb** [comment]
Hi @maicol07,

using middleware alias for the registration of the `OIDCStateMiddleware` class does not work.
I had tested before, but maybe I've missed something with that way to do.
The problem is that this middleware must be called between the `StartSession` and the `VerifyCsrfToken`  middlewares.
It rebuilds the session token only if the `state` parameter is set in the request, otherwise it does nothing.
For the moment, I don't see an obvious alternative or other paths to explore.
Any ideas?
Ok for the config setting.

**Comment by maicol07** [comment]
That's fine. I've implemented the custom state in the Client: [https://github.com/maicol07/oidc-client-php/blob/11ee3e91cf26113c28c8b475737bcc007afa99e4/src/Client.php#L320](https://github.com/maicol07/oidc-client-php/blob/11ee3e91cf26113c28c8b475737bcc007afa99e4/src/Client.php#L320)

Can you try that and see if it helps?

**Comment by eviweb** [comment]
@maicol07,
I found a way to simply enable the `OIDCStateMiddleware` only for the `callback` route and only when it is called using the `POST` method.
In the service provider register the middleware when the check below is `true`.
```php
 private function shouldEnableOIDCStateMiddleware(): bool
    {
        $request = \Illuminate\Http\Request::capture();
        $callbackPathInfo = '/' . config('oidc.provider_name') . '/' . config('oidc.callback_route_path');

        return 'POST' === strtoupper($request->method()) && $request->getPathInfo() === $callbackPathInfo;
    }
```
Is it convenient for you ?
From the user point of view I believe it is better to set this behaviour by default.
If you still want to give him/her the ability to enable/disable it, I can add a dedicated configuration setting, but I suggest that it should be used to disable the middleware registration instead.
Please tell me if you agree with this proposal.

**Comment by maicol07** [comment]
Yes, I agree. I'm not a fan of the yoda check style, but that isn't really important :sweat\_smile: 

The setting should be enabled by default and disabled if necessary

**Comment by eviweb** [comment]
@maicol07,
the PR is ready for review.

File: config/oidc.php [hunk 1]
```diff
@@ -31,5 +31,6 @@
 
     'callback_route_path' => env('OIDC_CALLBACK_ROUTE_PATH', 'callback'),
     'redirect_path_after_login' => env('OIDC_REDIRECT_PATH_AFTER_LOGIN', '/'),
-    'redirect_path_after_logout' => env('OIDC_REDIRECT_PATH_AFTER_LOGOUT', '/')
+    'redirect_path_after_logout' => env('OIDC_REDIRECT_PATH_AFTER_LOGOUT', '/'),
+    'disable_oidcstatemiddleware_for_post_callback' => false,
```
Comment (maicol07) [inline config/oidc.php]: You should add an env variable since this is a boolean

**Review by maicol07 — CHANGES_REQUESTED** [review]
Hi,
Sorry for the long wait, I've forgotten about this. Only one thing to fix and then the PR is ready to merge!

File: config/oidc.php (hunk 1 above)
Comment (eviweb) [reply config/oidc.php]: Hi @maicol07,
I let you review the last commit.
Thanks

**Review by maicol07 — CHANGES_REQUESTED** [review]
Ok, just a final request: instead of `oidcstatemiddleware` just call it `state_middleware` (with the underscore to divide words). This in the config setting and in the env. Also, this way the `oidc` word is not repeated in the ENV variable

**Comment by eviweb** [comment]
Hi @maicol07,
done. I don't think we came that close to merging :D

**Review by maicol07 — APPROVED** [review]

**Comment by maicol07** [comment]
Great! Merged!

**Comment by eviweb** [comment]
Thanks!
Well done for my oversight in the service provider ;).
Sorry for that.

**Pull request merged**